client/state/state_manager.js

```diff
@@ -13,86 +13,107 @@ See the License for the specific language governing permissions and
-    this.network_ = network;
+function isClosedOrStale(state) {
+  const now = time.now();
+  return (state.clientClosedTime < now - 5000) ||
+      (state.serverClosedTime < now - 5000) ||
+      (state.lastUpdatedTime < now - 600000); // 10 minutes.
+}
 
-    // Listen for communication from the server. NOTE: we rely on the network's
-    // own cleanup protocol to stop listening here.
-    // TODO(applmak): Vet that this works as expected and that no further
-    // disposal is needed.
-    network.on('newstate', (newstate) => {
-      if (newstate.name in this.trackedState_) {
-        var currentState = this.trackedState_[newstate.name];
-        if (currentState.owner_ === newstate.owner) {
-          return;
-        } else if (currentState.owner_ === undefined &&
-            network.id === newstate.owner) {
-          debug('Updating owner of owned state: ' + newstate.name);
-          currentState.owner_ = newstate.owner;
-          return;
+// A map of module id -> {
+//   state: {state name -> SharedState},
+//   clientClosedTime: timestamp,
+//   serverClosedTime: timestamp,
+// };
+const stateMap = {};
+export function forModule(network, id) {
+  return {
+    open() {
+      // Before we add another state, reap old ones.
+      for (const id in stateMap) {
+        // If the client closed this more than 5 seconds ago,
+        if (isClosedOrStale(stateMap[id])) {
+          delete stateMap[id];
         }
-        throw new Error('Received registration for state already owned by a ' +
-            'different client. This is likely because this client attempted ' +
-            'to create state with the same name as one already registered.');
       }
-      debug('Received new state registration ' + newstate.name);
-      this.trackedState_[newstate.name] = new sharedState.SharedState(
-          newstate.name,
-          sharedState.decodeInterpolator(newstate.interpolatorDef),
-          newstate.owner);
-    });
 
-    network.on('state', data => {
-      data.forEach(state => {
-        if (!(state.name in this.trackedState_)) {
-          debug('Data received for state that wasn\'t created! ' + state.name);
-          return;
-        }
-        if (state.dataPoint !== undefined) {
-          this.get(state.name).set(state.dataPoint.value, state.dataPoint.time);
+      if (!stateMap[id]) {
+        stateMap[id] = initState();
+      }
+      return {
+        define(stateName, def) {
+          assert(!(stateName in stateMap[id].state), `State ${stateName} was already defined!`);
+          stateMap[id].state[stateName] =
+              new SharedState(stateName, decodeInterpolator(def));
+          if (stateMap[id].priorData[stateName]) {
+            // We have some data that the server sent before we were ready.
+            // Add it to the shared state now.
+            for (const {data, time} of stateMap[id].priorData[stateName]) {
+              stateMap[id].state[stateName].set(data, time);
```

Comment:
Might help some module writer's sanity to log if we clobber a bit of state with priorState here, i.e. if the data are different but time is the same.

Reply:
I'm not 100% sure how to do that, so I added a TODO.